File: 315.cpp
/**
Online Judge : UVA Online Judge
Problem No   : 315
Problem Name : Network
Author       : Zubayer Rahman
Email        : zubayer.csesust@gmail.com
CPU          : 0.000s
Memory       :
Algorithm    : Articulation Point/Bridge
*/

#include<bits/stdc++.h>

#define MAX_NODES 100

using namespace std;

int n;

vector<int> vec[MAX_NODES+5];
set<int> s;

int visited[MAX_NODES],dis[MAX_NODES],low[MAX_NODES],parent[MAX_NODES];

int tim;

void DFS(int root)
{
    visited[root] = 1;
    dis[root]     = ++tim;
    low[root]     = dis[root];

    int v,child=0;
    int len = vec[root].size();

    while(!vec[root].empty())
    {
        v = *vec[root].begin();
        vec[root].erase(vec[root].begin());

        if(!visited[v])
        {
            child++;
            parent[v] = root;

            DFS(v);

            low[root] = min(low[root], low[v]);

            if(dis[root] <= low[v] && parent[root] != -1)
                s.insert(root);

            if(child>1 && parent[root]==-1)
                s.insert(root);
        }
        else if(parent[root] != v)
            low[root] = min(low[root], dis[v]);
    }
}

int main()
{
#ifndef ONLINE_JUDGE
    freopen("input.txt", "r", stdin);
#endif // ONLINE_JUDGE

    while(scanf("%d", &n)==1)
    {
        getchar();

        if(n==0)
            return 0;

        char str[1000];

        char *point;
        int u,v;

        while(scanf("%[^\n]", str) && strcmp(str, "0"))
        {
            getchar();

            point = strtok(str, " ");
            u = atoi(point);

            point = strtok(NULL, " ");

            while(point != NULL)
            {
                v = atoi(point);

                vec[u].push_back(v);
                vec[v].push_back(u);

                point = strtok(NULL, " ");
            }
        }

        memset(visited,0,sizeof(visited));
        memset(dis,0,sizeof(dis));
        memset(low,0,sizeof(low));
        memset(parent,-1,sizeof(parent));

        s.clear();

        tim=0;

        DFS(1);

        cout<<s.size()<<endl;
    }
    return 0;
}

File: 315/315.py
# Online Judge : UVA Online Judge
# Problem No   : 315
# Problem Name : Network
# Author       : Zubayer Rahman
# Email        : zubayer.csesust@gmail.com
# Time Limit   : 3.000s
# CPU          : 0.070s
# Memory       :
# Algorithm    : Articulation Point/Bridge


import sys
from collections import defaultdict


class Graph:
    MAX_NODES = 100

    def __init__(self):
        self.graph = defaultdict(list)

        self.visited = [False] * Graph.MAX_NODES
        self.discover_time = [0] * Graph.MAX_NODES
        self.lower = [0] * Graph.MAX_NODES
        self.parent = [-1] * Graph.MAX_NODES

        self.articulation_points = set()

        self.time = 0

    def addEdge(self, u: int, v: int):
        self.graph[u].append(v)
        self.graph[v].append(u)

    def DFS(self, root: int):
        self.time += 1

        self.visited[root] = True
        self.discover_time[root] = self.time
        self.lower[root] = self.discover_time[root]

        child: int = 0

        while self.graph[root]:
            v: int = self.graph[root].pop(0)

            if not self.visited[v]:
                child += 1
                self.parent[v] = root

                self.DFS(v)

                self.lower[root] = min(self.lower[root], self.lower[v])

                if self.discover_time[root] <= self.lower[v] and not self.parent[root] == -1:
                    self.articulation_points.add(root)

                if child > 1 and self.parent[root] == -1:
                    self.articulation_points.add(root)
            elif self.parent[root] != v:
                self.lower[root] = min(self.lower[root], self.discover_time[v])


def main():
    try:
        sys.stdin = open("input.txt", "r")
    except:
        pass

    while True:
        n = int(sys.stdin.readline().strip())
        if n == 0:
            break

        graph = Graph()

        while True:
            s = sys.stdin.readline().strip()
            if s == "0":
                break

            nodes = list(map(lambda x: int(x), s.split()))

            u = nodes[0]
            for v in nodes[1:]:
                graph.addEdge(u, v)
                graph.addEdge(v, u)

        graph.DFS(1)

        sys.stdout.write("{}\n".format(len(graph.articulation_points)))


if __name__ == '__main__':
    main()
